sdks/python/apache_beam/metrics/cells_test.py

```diff
@@ -0,0 +1,138 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import threading
+import unittest
+
+from apache_beam.metrics.cells import CounterCell
+from apache_beam.metrics.cells import DistributionCell
+from apache_beam.metrics.cells import DistributionData
+from apache_beam.metrics.cells import CellCommitState
+
+
+class TestCounterCell(unittest.TestCase):
+  @classmethod
+  def _modify_counter(cls, d):
+    for i in range(cls.NUM_ITERATIONS):
+      d.inc(i)
+
+  NUM_THREADS = 5
+  NUM_ITERATIONS = 100
+
+  def test_parallel_access(self):
+    # We create NUM_THREADS threads that concurrently modify the counter.
+    threads = []
+    c = CounterCell()
+    for _ in range(TestCounterCell.NUM_THREADS):
+      t = threading.Thread(target=TestCounterCell._modify_counter,
+                           args=(c,))
+      threads.append(t)
+      t.start()
+
+    for t in threads:
+      t.join()
+
+    total = (self.NUM_ITERATIONS * (self.NUM_ITERATIONS-1)/2 * self.NUM_THREADS)
+    self.assertEqual(c.get_cumulative(), total)
+
+  def test_basic_operations(self):
+    c = CounterCell()
+    c.inc(2)
+    self.assertEqual(c.get_cumulative(), 2)
+
+    c.dec(10)
+    self.assertEqual(c.get_cumulative(), -8)
+
+    c.dec()
+    self.assertEqual(c.get_cumulative(), -9)
+
+    c.inc()
+    self.assertEqual(c.get_cumulative(), -8)
+
+
+class TestDistributionCell(unittest.TestCase):
+  @classmethod
+  def _modify_distribution(cls, d):
+    for i in range(cls.NUM_ITERATIONS):
+      d.update(i)
+
+  NUM_THREADS = 5
+  NUM_ITERATIONS = 100
+
+  def test_parallel_access(self):
+    # We create NUM_THREADS threads that concurrently modify the distribution.
+    threads = []
+    d = DistributionCell()
+    for _ in range(TestDistributionCell.NUM_THREADS):
+      t = threading.Thread(target=TestDistributionCell._modify_distribution,
+                           args=(d,))
+      threads.append(t)
+      t.start()
+
+    for t in threads:
+      t.join()
+
+    total = (self.NUM_ITERATIONS * (self.NUM_ITERATIONS-1)/2 * self.NUM_THREADS)
+
+    count = (self.NUM_ITERATIONS * self.NUM_THREADS)
+
+    self.assertEqual(d.get_cumulative(),
+                     DistributionData(total, count, 0,
+                                      self.NUM_ITERATIONS - 1))
+
+  def test_basic_operations(self):
+    d = DistributionCell()
+    d.update(10)
+    self.assertEqual(d.get_cumulative(),
+                     DistributionData(10, 1, 10, 10))
+    self.assertEqual(d.get_cumulative().mean, 10)
+
+    d.update(2)
+    self.assertEqual(d.get_cumulative(),
+                     DistributionData(12, 2, 2, 10))
+    self.assertEqual(d.get_cumulative().mean, 6)
+
+    d.update(900)
+    self.assertEqual(d.get_cumulative(),
+                     DistributionData(912, 3, 2, 900))
+    self.assertEqual(d.get_cumulative().mean, 912/3)
+
```

Comment:
Can you test that passing float values to DistributionData will clip the float values to be integers?  For example, combining [3.1, 3.2, 3.3] here give would clip each value to int(3) first before accumulating into the distribution.  Also maybe test that passing float('inf') raises an OverflowError.

@bjchambers: since we don't have explicit typing in this API, does this seem reasonable to you?  Do we want to keep the int restriction on values in the distribution metric?